File: backend/app/models/inventory_accounting.py
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


InventoryMovementType = Literal[
    "purchase_receipt",
    "sales_issue",
    "sales_return",
    "purchase_return",
    "adjustment_in",
    "adjustment_out",
]


class InventoryMovementCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    sku_id: str = Field(min_length=1, max_length=64)
    warehouse_id: str = Field(min_length=1, max_length=64)
    movement_date: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")
    movement_type: InventoryMovementType
    quantity: Decimal = Field(gt=Decimal("0"), max_digits=16, decimal_places=4)
    amount: Decimal = Field(ge=Decimal("0"), max_digits=16, decimal_places=2)
    source_id: str = Field(min_length=1, max_length=120)


class InventoryMovement(InventoryMovementCreate):
    movement_id: str
    unit_cost: Decimal = Field(ge=Decimal("0"), max_digits=16, decimal_places=4)
    journal_entry_id: str | None = None


class InventoryBalance(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str
    sku_id: str
    warehouse_id: str
    quantity: Decimal = Field(max_digits=16, decimal_places=4)
    amount: Decimal = Field(max_digits=16, decimal_places=2)
    moving_average_cost: Decimal = Field(ge=Decimal("0"), max_digits=16, decimal_places=4)


class InventorySalesIssueResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str
    sku_id: str
    warehouse_id: str
    period: str
    movement_id: str
    source_id: str
    quantity: Decimal = Field(gt=Decimal("0"), max_digits=16, decimal_places=4)
    cost_amount: Decimal = Field(ge=Decimal("0"), max_digits=16, decimal_places=2)
    unit_cost: Decimal = Field(ge=Decimal("0"), max_digits=16, decimal_places=4)
    cogs_account_code: str = "6401"
    inventory_account_code: str = "1405"
    journal_entry_id: str

File: frontend/src/types/inventoryAccounting.ts
import type { JournalEntryRecord } from "./accounting";

export type MoneyValue = number | string;
export type QuantityValue = number | string;

export type InventoryMovementType =
  | "purchase_receipt"
  | "sales_issue"
  | "sales_return"
  | "purchase_return"
  | "adjustment_in"
  | "adjustment_out";

export interface InventoryMovement {
  account_set_id: string;
  sku_id: string;
  warehouse_id: string;
  movement_date: string;
  movement_type: InventoryMovementType;
  quantity: QuantityValue;
  amount: MoneyValue;
  source_id: string;
  movement_id: string;
  unit_cost: MoneyValue;
  journal_entry_id?: string | null;
}

export interface InventoryBalance {
  account_set_id: string;
  sku_id: string;
  warehouse_id: string;
  quantity: QuantityValue;
  amount: MoneyValue;
  moving_average_cost: MoneyValue;
}

export interface InventoryAccountingSummary {
  account_set_id: string;
  total_balances: number;
  total_movements: number;
  balances: InventoryBalance[];
  movements: InventoryMovement[];
}

export interface InventoryPurchaseReceiptRequest {
  account_set_id: string;
  sku_id: string;
  warehouse_id: string;
  period: string;
  quantity: QuantityValue;
  amount: MoneyValue;
  supplier_id: string;
}

export interface InventorySalesIssueRequest {
  account_set_id: string;
  sku_id: string;
  warehouse_id: string;
  period: string;
  quantity: QuantityValue;
}

export interface InventorySalesIssueResult {
  account_set_id: string;
  sku_id: string;
  warehouse_id: string;
  period: string;
  movement_id: string;
  source_id: string;
  quantity: QuantityValue;
  cost_amount: MoneyValue;
  unit_cost: MoneyValue;
  cogs_account_code: string;
  inventory_account_code: string;
  journal_entry_id: string;
}

export interface InventoryImpairmentRequest {
  account_set_id: string;
  sku_id: string;
  period: string;
  amount: MoneyValue;
}

export interface InventoryCountVarianceRequest {
  account_set_id: string;
  sku_id: string;
  warehouse_id: string;
  period: string;
  actual_quantity: QuantityValue;
  approved_by: string;
  approved_at: string;
}

export interface InventoryCountVarianceResult {
  account_set_id: string;
  sku_id: string;
  warehouse_id: string;
  period: string;
  variance_type: "gain" | "loss" | "none";
  book_quantity: QuantityValue;
  actual_quantity: QuantityValue;
  variance_quantity: QuantityValue;
  variance_amount: MoneyValue;
  source_id: string;
  approved_by: string;
  approved_at: string;
  journal_entry_id?: string | null;
}

export type InventoryImpairmentResult = JournalEntryRecord;
